mathics/builtin/patterns/rules.py

```diff
@@ -0,0 +1,613 @@
+# -*- coding: utf-8 -*-
+"""
+Define, apply and compiling rules.
+"""
+
+from typing import List, Optional as OptionalType, Tuple, Union
+
+from mathics.core.atoms import Integer, Integer2, Number, String
+from mathics.core.attributes import A_HOLD_REST, A_PROTECTED, A_SEQUENCE_HOLD
+from mathics.core.builtin import AtomBuiltin, BinaryOperator, Builtin, PatternError
+from mathics.core.element import BaseElement
+from mathics.core.evaluation import Evaluation
+from mathics.core.exceptions import InvalidLevelspecError
+from mathics.core.expression import Expression
+from mathics.core.list import ListExpression
+from mathics.core.rules import Rule
+from mathics.core.symbols import Atom, Symbol, SymbolList, SymbolTrue
+from mathics.core.systemsymbols import (
+    SymbolDispatch,
+    SymbolInfinity,
+    SymbolRule,
+    SymbolRuleDelayed,
+)
+from mathics.eval.parts import python_levelspec
+
+# This tells documentation how to sort this module
+sort_order = "mathics.builtin.rules-and-patterns.rules"
+
+
+class Rule_(BinaryOperator):
+    """
+
+    <url>:WMA link:https://reference.wolfram.com/language/ref/Rule_.html</url>
+
+    <dl>
+      <dt>'Rule[$x$, $y$]'
+      <dt>'$x$ -> $y$'
+      <dd>represents a rule replacing $x$ with $y$.
+    </dl>
+
+    >> a+b+c /. c->d
+    = a + b + d
+    >> {x,x^2,y} /. x->3
+     = {3, 9, y}
+    >> a /. Rule[1, 2, 3] -> t
+     : Rule called with 3 arguments; 2 arguments are expected.
+     = a
+    """
+
+    name = "Rule"
+    operator = "->"
+    attributes = A_SEQUENCE_HOLD | A_PROTECTED
+    grouping = "Right"
+    needs_verbatim = True
+    summary_text = "a replacement rule"
+
+    def eval_rule(self, elems, evaluation):
+        """Rule[elems___]"""
+        num_parms = len(elems.get_sequence())
+        if num_parms != 2:
+            evaluation.message("Rule", "argrx", "Rule", Integer(num_parms), Integer2)
+        return None
+
+
+class RuleDelayed(BinaryOperator):
+    """
+    <url>:WMA link:https://reference.wolfram.com/language/ref/RuleDelayed.html</url>
+
+    <dl>
+      <dt>'RuleDelayed[$x$, $y$]'
+      <dt>'$x$ :> $y$'
+      <dd>represents a rule replacing $x$ with $y$, with $y$ held
+        unevaluated.
+    </dl>
+
+    >> Attributes[RuleDelayed]
+     = {HoldRest, Protected, SequenceHold}
+    """
+
+    attributes = A_SEQUENCE_HOLD | A_HOLD_REST | A_PROTECTED
+    needs_verbatim = True
+    operator = ":>"
+    summary_text = "a rule that keeps the replacement unevaluated"
+
+    def eval_rule_delayed(self, elems, evaluation):
+        """RuleDelayed[elems___]"""
+        num_parms = len(elems.get_sequence())
+        if num_parms != 2:
+            evaluation.message(
+                "RuleDelayed", "argrx", "RuleDelayed", Integer(num_parms), Integer2
+            )
+        return None
+
+
+# TODO: disentangle me
+def create_rules(
```

Comment:
This is really an eval function and should be moved to `mathics.eval.rule` and probably called `eval_Replace()`.

Reply:
Will be addressed in PR.